### DISCUSSION TITLE
Forecasting time step configuration

### DISCUSSION THREAD

**Discussion by shamshaw** [neuralhydrology/neuralhydrology#89] [Q&A] [answered]
Hello,

I'm interested to just clarify some of the configuration parameters that would be set to perform forecasting experiments (predicting streamflow at a future time step). Am I correct in understanding the terminology that if `predict_last_n = 1`, then the model would be predicting streamflow at time _t0_? 

So, if you wanted to take the last 100 days of forcing input observations and predict streamflow 10 days ahead, then am I correct that the settings would be a `seq_len = 110` and `predict_last_n = 10`? I would then have streamflow predictions from t+1 to t+10.

Of course, alternatively if I was interested in just the streamflow at t+10 I could create a target variable of the 10 day ahead streamflow and use that, but using the config settings would be more flexible so just want to make sure I understand the correct indexing of time steps.

Thanks!

**Comment by kratzert** [comment, marked as answer]
Hi @shamshaw 

sorry, I somehow missed to answer your questions. There are a couple of ways how you could do this experiment but to be honest, I would probably implement a dedicated "Forecast LSTM" for this, as e.g. in [this abstract (includes a video)](https://meetingorganizer.copernicus.org/EGU21/EGU21-9714.html) or [this paper (see figure 2b)](https://hess.copernicus.org/preprints/hess-2021-554/#discussion). 

Regarding your specific question: `seq_len: 110` and `predict_last_n: 10` would not result in a 10 day forecast starting at time step 100. Instead, the LSTM would use the full 110 timesteps and the model output of the last 10 timesteps would be considered in the loss. This means the model will also use the forcings for the last 10 timesteps, i.e. not really "forecasting" 10 timesteps into "the future". 

The easiest thing that comes to my mind that would not require any code changes would be the following. Using the `duplicate_features` config argument, paired with `lagged_features`, I would create 10 copies of the target feature and then shift these ten copies by 1, 2, 3, ... 10 days. You could then set all 10 of these shifted copies as `target_variables`, which means that at every timestep, the model predict 10 values that are fitted against the different shifted targets. If you now set `predict_last_n: 1` and `seq_len: 100` your model would use an input sequence of 100 timesteps and at the last timestep, it produces 10 outputs that are fitted against q(t+1), q(t+2), ..., q(t+10).

However, I think (from my own experience) it makes more sense to implement a encoder-decoder LSTM structure, where the encoder processes the 100 timesteps of meteorological inputs and the decoder structure then runs for n timesteps (n=10 in your case) and at every timestep produces the output just of that day. In this setting, you could also provide features available during forecast (static attributes and weather forecast) to the decoder model. This will however involve a bit of programming, but I think a good starting point is the MTS-LSTM class, which already includes a handoff LSTM structure. If that code is to intimidating (because it includes a lot of overhead for handling the frequency stuff) then just start from scratch (or e.g. the cudalstm) and add a second LSTM cell to this model that is initialized with the hidden/cell state of the last time step of the encoder and then produces the forecast.

Is my explanation clear? Otherwise feel free to ask again.

**Reply by shamshaw** [reply]
Thanks @kratzert ! Your explanation makes sense and that helps clear up how the `predict_last_n` parameter is used. I agree with you that if we wanted to forecast a time series into the future, using a different algorithm like an encoder-decoder LSTM would make sense. I could see for some initial modeling, it would be easy to create a single future target (like predicting just t+10) using the built in models but for more robust forecasting, we want to develop something specific for that use case. Thanks for the links and explanation!
